// llrpjs/dist/src/proxy/parameter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LLRPProxyParameter = void 0;
const parameter_1 = require("../element/parameter");
const buffer_1 = require("../buffer/buffer");
class LLRPProxyParameter {
    constructor(args) {
        if (args instanceof Buffer) {
            this.origin = new parameter_1.LLRPParameter(new buffer_1.LLRPBuffer(args));
        }
        else if (args instanceof parameter_1.LLRPParameter) {
            this.origin = args;
        }
        else {
            this.origin = new parameter_1.LLRPParameter(args);
        }
    }
    getName() {
        return this.origin.getName();
    }
    encode() {
        this.origin.assemble().encode();
        return this;
    }
    decode() {
        this.origin.decode().marshal();
        return this;
    }
    toLLRPData() {
        return this.origin.toLLRPData();
    }
    getBuffer() {
        return this.origin.getBuffer().getBuffer();
    }
}
exports.LLRPProxyParameter = LLRPProxyParameter;
