// llrpjs/dist/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LLRPServer = exports.LLRPClient = exports.LLRPParameter = exports.LLRPMessage = exports.LLRPCore = exports.LLRPCoreDef = void 0;
const def_1 = require("./def");
Object.defineProperty(exports, "LLRPCoreDef", { enumerable: true, get: function () { return def_1.LLRPDef; } });
const message_1 = require("./typed/message");
const parameter_1 = require("./typed/parameter");
const class_registry_1 = require("./registry/class-registry");
const type_registry_1 = require("./registry/type-registry");
const client_1 = require("./net/client");
const server_1 = require("./net/server");
type_registry_1.TypeRegistry.getInstance().enrollCoreDefinitions(def_1.LLRPDef).build();
const CR = class_registry_1.LLRPClassRegistry.getInstance(def_1.LLRPDef).enrollCoreDefinitions(def_1.LLRPDef).build();
const LLRPCoreMessages = CR.getCoreMessageClasses();
const LLRPCoreParameters = CR.getCoreParamClasses();
const LLRPCore = { ...LLRPCoreMessages, ...LLRPCoreParameters };
exports.LLRPCore = LLRPCore;
const LLRPMessage = message_1.LLRPTypedMessage.ofDef(def_1.LLRPDef);
exports.LLRPMessage = LLRPMessage;
const LLRPParameter = parameter_1.LLRPTypedParameter.ofDef(def_1.LLRPDef);
exports.LLRPParameter = LLRPParameter;
const LLRPClient = client_1.LLRPClientOfDef(def_1.LLRPDef);
exports.LLRPClient = LLRPClient;
const LLRPServer = server_1.LLRPServerOfDef(def_1.LLRPDef);
exports.LLRPServer = LLRPServer;
