// llrpjs/dist/src/element/header-fd.js
"use strict";
/**
 * Message header field descriptors
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LLRP_TD_TV_TYPENUM = exports.LLRP_TD_PARAM_LENGTH = exports.LLRP_TD_RSVD_TYPENUM = exports.LLRP_TD_MESSAGE_ID = exports.LLRP_TD_MESSAGE_LENGTH = exports.LLRP_TD_RSVD_VERSION_TYPENUM = void 0;
exports.LLRP_TD_RSVD_VERSION_TYPENUM = {
    name: "RsvdVersionTypeNum",
    type: "u16",
    format: "Normal"
};
exports.LLRP_TD_MESSAGE_LENGTH = {
    name: "MessageLength",
    type: "u32",
    format: "Normal"
};
exports.LLRP_TD_MESSAGE_ID = {
    name: "MessageID",
    type: "u32",
    format: "Normal"
};
/**
 * TLV parameter header field descriptors
 */
exports.LLRP_TD_RSVD_TYPENUM = {
    name: "RsvdTLVTypeNum",
    type: "u16",
    format: "Normal"
};
exports.LLRP_TD_PARAM_LENGTH = {
    name: "TLVparamLength",
    type: "u16",
    format: "Normal"
};
/**
 * TV parameter header field descriptor
 */
exports.LLRP_TD_TV_TYPENUM = {
    name: "TVTypeNum",
    type: "u8",
    format: "Normal"
};
